### PR TITLE
2.6 - Model Cache Returns Applications and Machines as Copies

### PR DESCRIPTION
**Pull request by manadart**
## Description of change

- Applications and machines are supplied by the cache as copies instead of references.
- Copy methods are implemented on `<Entity>Change` structs.

## QA steps

No materialised changes. New unit tests cover copying.
Regressions can be checked with the same steps as for #10062

## Documentation changes

None.

## Bug reference

N/A

### CODE REVIEW

**Comment by manadart** [comment]
[Implement Deep Copying for More Cached Entities](https://trello.com/c/vvEknEz6/1525-implement-deep-copying-for-more-cached-entities)

File: core/cache/application_test.go [hunk 1]
```diff
@@ -23,10 +23,17 @@ var _ = gc.Suite(&ApplicationSuite{})
...
 	m.Config()
 	c.Check(testutil.ToFloat64(s.Gauges.ApplicationConfigReads), gc.Equals, float64(1))
 	m.Config()
 	c.Check(testutil.ToFloat64(s.Gauges.ApplicationConfigReads), gc.Equals, float64(2))
+
+	// Goroutine safety.
+	go m.Config()
+	go m.Config()
+	go m.Config()
+	c.Check(testutil.ToFloat64(s.Gauges.ApplicationConfigReads), gc.Equals, float64(5))
```
Comment (achilleasa) [inline core/cache/application_test.go]: This looks flaky as you cannot guarantee that the go-routines will spin up and actually run before you hit the check (on a fast machine I suspect this test will fail). How about changing this to :

```go
var wg sync.WaitGroup
wg.Add(3)
for i:=0;i<3;i++{
  go func(){ 
     defer wg.Done()
     m.Config()
  }()
}
wg.Wait()
c.Check(...)
```

Comment (manadart) [reply core/cache/application_test.go]: Rookie mistake. Yes.

**Comment by manadart** [comment]
The mutation and the retrieval are both lock protected at the model level, so as long as this access pattern is preserved, such a race will not happen.

There is always the possibility that a retrieved object changes in the cache after it is acquired, but of course that is the same as with any cache implementation.

**Comment by manadart** [comment]
$$merge$$

**Pull request merged**